mogp_gpu/src/gpparams.hpp

```diff
@@ -0,0 +1,240 @@
+
+    void set_mean(vec _new_mean) {
+        mean = _new_mean;
+        n_mean = _new_mean.size();
+    }
+
+    inline vec get_mean() const { return mean; }
+
+    inline int get_n_mean() const { return n_mean; }  
+
+    inline vec get_data() const { return data; }
+
+    void set_data(vec _new_data) {
+        /// also resets Mean, covariance, nugget
+        if (_new_data.size() != n_data) 
+            throw std::runtime_error("New data not correct shape");
+        data = _new_data;
+        has_data = true;
+        mean = vec::Zero(n_mean);
+        if (nug_type == NUG_ADAPTIVE) nugget_size = 0.;
+    }
+
+    vec get_corr_raw() const {
+        return data.block(0, 0, n_corr, 1);
+    }
+
+    vec get_corr() const {
+        return CorrTransform().raw_to_scaled(get_corr_raw());
+    }
+
+    void set_corr(vec _new_corr) {
+        data.block(0,0,n_corr,1) = CorrTransform().scaled_to_raw(_new_corr);
+    }
+
+    inline int get_n_corr() { return n_corr; }
+
+    int get_cov_index() const {
+        // determine where in the data array the covariance parameter is
+        if (nug_type == NUG_FIT) return n_data - 2;
+        else return n_data - 1;
+    }
+
+    // transformed covariance parameter
+    REAL get_cov() const {
+       
+        if (! has_data) return 0.;
+        else return CovTransform().raw_to_scaled(data[get_cov_index()]);
+        
+    }
+
+    void set_cov(REAL _cov) {
+        if (! has_data) throw  std::runtime_error("Need to set data before setting covariance parameter");
+        data[get_cov_index()] = CovTransform().scaled_to_raw(_cov);
+    }
+
+    inline void set_nugget_type(nugget_type _nug_type) { nug_type = _nug_type; }
+
+    inline nugget_type get_nugget_type() const { return nug_type; }
+
+    void set_nugget_size(REAL _nug_size) {
+        nugget_size = _nug_size;
+        if ((nug_type == NUG_FIT) && (data.size()>0)) {
+            data[n_data -1] = nugget_size;
+        }
+    }
+
+    REAL get_nugget_size() const {    
+        if (nug_type != NUG_FIT) {
+            return nugget_size; 
+        } else {
+            return CovTransform().raw_to_scaled(data[n_data-1]);
+        }
+    }   
+
+    bool test_same_shape(GPParams& other) const {
+        if (n_mean != other.get_n_mean()) return false;
+        if (n_corr != other.get_n_corr()) return false;
+        if (nug_type != other.get_nugget_type()) return false;
+        return true;
+    }
+
+    bool test_same_shape(vec& other) const {
+        if ((get_n_mean() + get_n_data()) != other.size()) return false;
+        return true;
+    }
+
+    bool test_close(GPParams& other, REAL threshold=1e-8) const {
+        return (get_data() - other.get_data()).norm() < threshold;
+    }
+
+    bool test_close(vec& other, REAL threshold=1e-8) const {
+        return (get_data() - other).norm() < threshold;
+    }
+
+    void unset_data() { 
+        mean = vec::Zero(n_mean);
```

Comment:
No strong opinion on this, but I quite like having the flag, so it can be explicitly set if and only if the parameters are fit successfully.   (Also it looks like the Eigen syntax for initializing vectors to NaN might be a bit long-winded!)

Reply:
Ah, I see. I think as long as you check the flag when predicting and raising and error if the parameters weren't set then I don't think there's a problem here. You might also consider using some values that aren't realistic as a way to indicate that parameters aren't set as all zeros might be seen reasonable values to use!